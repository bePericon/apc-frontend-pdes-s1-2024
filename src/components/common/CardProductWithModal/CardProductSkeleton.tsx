import { Skeleton } from '@mui/material'
import { StyledSkeletonContainer, StyledSkeletonInfoContainer } from './Modal.styled'
import { useWidth } from '@/hook/useWidth'

const CardProductSkeleton = ({
    showCommentSection,
}: {
    showCommentSection?: boolean
}) => {
    const { isMobile } = useWidth()

    return (
        <StyledSkeletonContainer>
            <Skeleton
                variant="rectangular"
                width={isMobile ? 300 : 400}
                height={isMobile ? 240 : 400}
            />

            <StyledSkeletonInfoContainer>
                <Skeleton variant="text" sx={{ fontSize: '2.5rem', width: '100%' }} />
                <Skeleton variant="text" sx={{ fontSize: '2.5rem', width: '70%' }} />

                <div
                    style={{
                        display: 'flex',
                        flexDirection: 'row',
                        justifyContent: 'space-between',
                        alignItems: 'flex-start',
                        width: '100%',
                    }}
                >
                    <Skeleton variant="text" sx={{ fontSize: '1.5rem', width: '30%' }} />
                    <Skeleton variant="text" sx={{ fontSize: '1.5rem', width: '10%' }} />
                </div>
                <Skeleton variant="text" sx={{ fontSize: '1.5rem', width: '30%' }} />
                <Skeleton variant="text" sx={{ fontSize: '1.5rem', width: '60%' }} />
                <Skeleton variant="text" sx={{ fontSize: '1.2rem', width: '70%' }} />

                <div
                    style={{
                        display: 'flex',
                        flexDirection: 'row',
                        justifyContent: 'flex-end',
                        alignItems: 'flex-end',
                        gap: 16,
                        width: '100%',
                    }}
                >
                    <Skeleton variant="text" sx={{ fontSize: '1.5rem', width: '25%' }} />
                </div>

                {showCommentSection && (
                    <>
                        <Skeleton variant="rectangular" width={'100%'} height={120} />
                        <div
                            style={{
                                display: 'flex',
                                flexDirection: 'row',
                                justifyContent: 'center',
                                alignItems: 'flex-start',
                                gap: 16,
                                width: '100%',
                            }}
                        >
                            <Skeleton
                                variant="text"
                                sx={{ fontSize: '1.5rem', width: '10%' }}
                            />
                            <Skeleton
                                variant="text"
                                sx={{ fontSize: '1.5rem', width: '10%' }}
                            />
                        </div>
                    </>
                )}
            </StyledSkeletonInfoContainer>
        </StyledSkeletonContainer>
    )
}

export default CardProductSkeleton
