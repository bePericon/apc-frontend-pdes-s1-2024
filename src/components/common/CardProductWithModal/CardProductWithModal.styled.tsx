import styled, { css } from 'styled-components'

export const StyledCategoryCard = styled.div`
    width: 100%;
    max-width: auto;
    min-height: 72px;
    display: flex;
    align-items: center;
    padding: 8px;
    border-radius: 8px;
    box-shadow: ${({ theme }) => theme.boxShadows.whiteS};
    background-color: #fff;
    cursor: pointer;

    &:hover {
        background-color: #faf0ca;
        outline: none;
    }

    &.focused {
        background-color: #fff;
        outline: none;
    }

    ${({ theme }) => {
        return css`
            @media screen and (max-width: ${theme.breakpoints.xxs}) {
                justify-content: center;
                width: 90%;
                height: auto;
                min-height: 52px;
                padding: 8px;
            }
        `
    }}
`

export const StyledCategoryCardInner = styled.div`
    display: flex;
`

export const StyledButton = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    cursor: pointer;
`

export const StyledInfo = styled.div`
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    gap: 4px;
    min-width: 400px;
    max-width: 400px;
    padding-top: 4px;

    [role='heading'] {
        line-height: 1.43;
        letter-spacing: normal;
    }

    ${({ theme }) => {
        return css`
            @media screen and (max-width: ${theme.breakpoints.xxs}) {
                min-height: auto;
            }

            @media screen and (max-width: ${theme.breakpoints.xs}) {
                min-width: 100px;
                max-width: 200px;
            }
        `
    }}
`

export const StyledFirstIcon = styled.div`
    align-self: flex-start;
    margin-right: 8px;
`

export const StyledRenderIcon = styled.div`
    align-self: flex-start;
`

export const StyledLastIcon = styled.div`
    align-self: center;
`
