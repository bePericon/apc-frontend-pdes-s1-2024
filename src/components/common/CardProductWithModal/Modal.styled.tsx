import { Typography } from '@mui/material'
import styled, { css } from 'styled-components'

export const StyledTitleContainer = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
`

export const StyledTypographyTitle = styled(Typography)`
    font-size: 2rem !important;

    ${({ theme }) => {
        return css`
            @media screen and (max-width: ${theme.breakpoints.xxs}) {
                font-size: 1.1rem !important;
                line-height: 1.1 !important;
            }
        `
    }}
`

export const StyledSkeletonContainer = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: flex-start;
    align-items: flex-start;
    gap: 32px;

    ${({ theme }) => {
        return css`
            @media screen and (max-width: ${theme.breakpoints.xxs}) {
                flex-direction: column;
            }
        `
    }}
`

export const StyledContainer = styled.div`
    display: flex;
    flex-direction: row;
    gap: 24px;

    ${({ theme }) => {
        return css`
            @media screen and (max-width: ${theme.breakpoints.xxs}) {
                flex-direction: column;
            }
        `
    }}
`

export const StyledInfoContainer = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
    gap: 16px;
`

export const StyledInnerContainer = styled.div`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
    align-items: center;
    width: 100%;
`

export const StyledSkeletonInfoContainer = styled.div`
    width: 50%;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    align-items: flex-start;
    gap: 4px;

    ${({ theme }) => {
        return css`
            @media screen and (max-width: ${theme.breakpoints.xxs}) {
                width: 100%;
            }
        `
    }}
`

export const StyledPurchaseSection = styled.div`
    display: flex;
    flex-direction: column;
    width: 100%;
    margin-top: 16px;
`

export const StyledPurchaseInnerSection = styled.div`
    display: flex;
    align-items: center;
    gap: 16px;
`

export const StyledPurchaseButtonContainer = styled.div`
    display: flex;
    margin-top: 8px;
    justify-content: flex-end;
`
