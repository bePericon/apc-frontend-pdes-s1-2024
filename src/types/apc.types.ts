export interface Role {
  _id: string;
  name: string;
  description: string;
  permissions: [
    {
      _id: string;
      name: string;
      description: string;
      creationDate: string;
    }
  ];
  creationDate: string;
}

export interface Favorite {
  _id: string;
  user: string;
  itemId: string;
  comment: string;
  rating: number;
  creationDate: string;
}

export interface User {
  _id: string;
  name: string;
  surname: string;
  username: string;
  email: string;
  password: string;
  roles: Role[];
  favorites: Favorite[];
  creationDate: string;
}
