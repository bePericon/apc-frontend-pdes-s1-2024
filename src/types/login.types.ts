export interface LoginData {
    email: string
    password: string
}

export interface LoginUserData {
    user: UserData
    // token: string;
}

export interface UserData {
    name: string
    surname: string
    username: string
    email: string
    createdDate?: Date
}

export interface SignUpData extends LoginData {
    name: string
}
