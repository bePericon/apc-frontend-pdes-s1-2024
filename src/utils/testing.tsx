import React, { PropsWithChildren } from "react";
import { render as rtlRender } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../redux/store";
import { theme } from "@/styles/theme";
import { DefaultTheme, ThemeProvider } from "styled-components";

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: Partial<RootState>;
  store?: AppStore;
}

// Wait for the promises to finish
const flushPromises = () => new Promise(process.nextTick);

function render(
  ui: React.ReactElement,
  extendedRenderOptions: ExtendedRenderOptions = {}
) {
  const {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  } = extendedRenderOptions;

  const Wrapper = ({ children }: PropsWithChildren) => (
    <Provider store={store}>
      <ThemeProvider theme={theme as DefaultTheme}>{children}</ThemeProvider>
    </Provider>
  );

  // Return an object with the store and all of RTL's query functions
  return {
    store,
    ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export * from "@testing-library/react";
export { render, flushPromises };
