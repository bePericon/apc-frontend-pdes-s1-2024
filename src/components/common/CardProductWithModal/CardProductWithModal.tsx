import {
    StyledButton,
    StyledCategoryCard,
    StyledFirstIcon,
    StyledInfo,
    StyledLastIcon,
} from './CardProductWithModal.styled'
import { ChevronRight } from '@mui/icons-material'
import Image from 'next/image'
import { Typography } from '@mui/material'
import { useState } from 'react'
import { Product } from '@/types/meli.types'

interface CardProductWithModalProps {
    item: Product
    setResearch?: () => void
    index: number
    renderModalComponent: (handleOnClose: () => void, open: boolean, item: Product) => void
}

const CardProductWithModal = ({
    item,
    setResearch,
    index,
    renderModalComponent,
}: CardProductWithModalProps) => {
    const [open, setOpen] = useState(false)

    const handleOnClick = async () => setOpen(true)

    const handleOnClose = () => {
        setOpen(false)
        if(setResearch) setResearch()
    }

    return (
        <>
            <StyledCategoryCard
                onClick={handleOnClick}
                data-test-id={`card-item-${index}`}
            >
                <StyledFirstIcon>
                    {item.hydrated?.thumbnail && (
                        <Image
                            src={item.hydrated?.thumbnail}
                            width={90}
                            height={90}
                            alt={item.hydrated?.thumbnail_id}
                            style={{ borderRadius: 8 }}
                        />
                    )}
                </StyledFirstIcon>
                <div>
                    <StyledInfo>
                        <Typography>{item.hydrated?.title}</Typography>
                    </StyledInfo>
                </div>
                <StyledLastIcon>
                    <StyledButton>
                        <ChevronRight />
                    </StyledButton>
                </StyledLastIcon>
            </StyledCategoryCard>
            {renderModalComponent(handleOnClose, open, item)}
        </>
    )
}

export default CardProductWithModal
