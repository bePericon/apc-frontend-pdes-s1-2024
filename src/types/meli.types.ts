import { User } from './apc.types'

export interface FavoriteProduct {
    favoriteId?: string
    rating?: number
    comment?: string
    isFavorite?: boolean
}

export interface Product extends FavoriteProduct {
    title: string
    thumbnail: string
    thumbnail_id: string
    pictures: any[]
    price: number
    itemId: string

    userId?: string
    createdDate?: string
    hydrated?: HydratedProduct
}

export interface PurchaseProduct extends Product {
    purchaseId: string
    price: number
    quantity: number
}

export interface HydratedProduct {
    itemId: string
    title: string
    thumbnail: string
    thumbnail_id: string
    pictures: Picture[]
    price: number
}

export interface Picture {
    quality: string
    id: string
    url: string
    secure_url: string
    size: string
    max_size: string
}
