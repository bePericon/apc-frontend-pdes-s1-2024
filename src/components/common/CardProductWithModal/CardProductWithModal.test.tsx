import { render } from '@/utils/testing'
import { Product } from '@/types/meli.types'
import '@testing-library/jest-dom'
import CardProductWithModal from './CardProductWithModal'

describe('CardItemList', () => {
    const product: Product = {
        favoriteId: '6657c4c295902dba19168195',
        createdDateFavorite: '2024-05-30T00:13:54.967Z',
        user: '6647867aa9551b91eb2e36e1',
        itemId: 'MLA1411220617',
        comment: '',
        rating: 9,
        hydrated: {
            itemId: 'MLA1411220617',
            title: 'Samsung Galaxy S22 Ultra 5g (snapdragon) 5g 512 Gb  Phantom Black 12 Gb Ram',
            thumbnail: 'http://http2.mlstatic.com/D_913503-MLA52132116989_102022-O.jpg',
            thumbnail_id: '913503-MLA52132116989_102022',
            pictures: [
                {
                    url: 'http://http2.mlstatic.com/D_913503-MLA52132116989_102022-O.jpg',
                    secure_url:
                        'https://http2.mlstatic.com/D_913503-MLA52132116989_102022-O.jpg',
                    size: '250x500',
                    max_size: '392x784',
                    quality: '',
                    id: '913503-MLA52132116989_102022',
                },
                {
                    size: '250x500',
                    max_size: '532x1064',
                    quality: '',
                    id: '982637-MLA52132138394_102022',
                    url: 'http://http2.mlstatic.com/D_982637-MLA52132138394_102022-O.jpg',
                    secure_url:
                        'https://http2.mlstatic.com/D_982637-MLA52132138394_102022-O.jpg',
                },
            ],
            price: 1849999,
        },
    }

    it('should render correctly', () => {
        const result = render(
            <CardProductWithModal
                item={product}
                index={0}
                renderModalComponent={() => {}}
            />
        )

        expect(result).toMatchSnapshot()
    })

    it('should show correct title', () => {
        const { getByText } = render(
            <CardProductWithModal
                item={product}
                index={0}
                renderModalComponent={() => {}}
            />
        )

        expect(getByText(/Samsung Galaxy/)).toBeInTheDocument()
    })
})
