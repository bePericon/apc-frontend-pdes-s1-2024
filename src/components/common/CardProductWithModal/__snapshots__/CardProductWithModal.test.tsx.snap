// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardItemList should render correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-8e39d0e7-0 eBdvIW"
        data-test-id="card-item-0"
      >
        <div
          class="sc-8e39d0e7-4 hbuvQA"
        >
          <img
            alt="913503-MLA52132116989_102022"
            data-nimg="1"
            decoding="async"
            height="90"
            loading="lazy"
            src="/_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_913503-MLA52132116989_102022-O.jpg&w=256&q=75"
            srcset="/_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_913503-MLA52132116989_102022-O.jpg&w=96&q=75 1x, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_913503-MLA52132116989_102022-O.jpg&w=256&q=75 2x"
            style="color: transparent; border-radius: 8px;"
            width="90"
          />
        </div>
        <div>
          <div
            class="sc-8e39d0e7-3 fvRaGh"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
            >
              Samsung Galaxy S22 Ultra 5g (snapdragon) 5g 512 Gb  Phantom Black 12 Gb Ram
            </p>
          </div>
        </div>
        <div
          class="sc-8e39d0e7-6 bTZkRN"
        >
          <div
            class="sc-8e39d0e7-2 ifgIgX"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="ChevronRightIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-8e39d0e7-0 eBdvIW"
      data-test-id="card-item-0"
    >
      <div
        class="sc-8e39d0e7-4 hbuvQA"
      >
        <img
          alt="913503-MLA52132116989_102022"
          data-nimg="1"
          decoding="async"
          height="90"
          loading="lazy"
          src="/_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_913503-MLA52132116989_102022-O.jpg&w=256&q=75"
          srcset="/_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_913503-MLA52132116989_102022-O.jpg&w=96&q=75 1x, /_next/image?url=http%3A%2F%2Fhttp2.mlstatic.com%2FD_913503-MLA52132116989_102022-O.jpg&w=256&q=75 2x"
          style="color: transparent; border-radius: 8px;"
          width="90"
        />
      </div>
      <div>
        <div
          class="sc-8e39d0e7-3 fvRaGh"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
          >
            Samsung Galaxy S22 Ultra 5g (snapdragon) 5g 512 Gb  Phantom Black 12 Gb Ram
          </p>
        </div>
      </div>
      <div
        class="sc-8e39d0e7-6 bTZkRN"
      >
        <div
          class="sc-8e39d0e7-2 ifgIgX"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ChevronRightIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6 8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
